File: Polynomial Regression/polynomial_regression.py
# Polynomial regression

# Import libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Import dataset
dataset = pd.read_csv('Position_Salaries.csv')

X = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2].values

# Linear regression model (for comparison)
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X, y)

# Polynomial regression model
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 4)
X_poly = poly_reg.fit_transform(X)

lin_reg_2 = LinearRegression()
lin_reg_2.fit(X_poly, y)

# Visualize linear regression
plt.scatter(X, y, color = "red")
plt.plot(X, lin_reg.predict(X), color = "blue")
plt.title("Linear model fit")
plt.xlabel("Position index")
plt.ylabel("Salary")
plt.show()

# Visualize polynomial regression
X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape((len(X_grid), 1))
plt.scatter(X, y, color = "red")
plt.plot(X, lin_reg_2.predict(X_poly), color = "blue")
plt.title("Polynomial model fit")
plt.xlabel("Position index")
plt.ylabel("Salary")
plt.show()

# Prediction with linear model
lin_reg.predict(6.5)

# Prediction with polynomial model
lin_reg_2.predict(poly_reg.fit_transform(6.5))

File: Polynomial Regression/polynomial_regression.R
# Importing the dataset
dataset <- read.csv('Position_Salaries.csv')
dataset <- dataset[2:3]

# Linear regression
lin_reg <- lm(Salary ~ Level, data = dataset)

# Polynomial regression
dataset$Level2 <- dataset$Level^2
dataset$Level3 <- dataset$Level^3
poly_reg <- lm(Salary ~ ., data = dataset)

# Visualising the Regression Model results
library(ggplot2)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') +
  geom_line(aes(x = dataset$Level, y = predict(lin_reg, newdata = dataset)),
            colour = 'blue') +
  ggtitle('Linear regression') +
  xlab('Level') +
  ylab('Salary')

# Visualising the polynomial model results (for higher resolution and smoother curve)
x_grid <- seq(min(dataset$Level), max(dataset$Level), 0.1)
x_poly <- data.frame(Level = x_grid, Level2 = x_grid^2, Level3 = x_grid^3)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') +
  geom_line(aes(x = x_grid, y = predict(poly_reg, newdata = x_poly)),
            colour = 'blue') +
  ggtitle('Polynomial model') +
  xlab('Level') +
  ylab('Salary')

# Linear prediction
y_pred <- predict(lin_reg, data.frame(Level = 6.5))

# Polynomial prediction
y_pred_poly <- predict(poly_reg, data.frame(Level = 6.5, Level2 = 6.5^2, Level3 = 6.5^3))
